react component for buttons

trishla08 opened the pull request:
buttonApp.jsx has been added as the required react component. as per https://github.com/julianguyen/ifme/issues/777

Edits are allowed from maintainers.

julianguyen commented:
Hey @trishla08, thanks for working on this :D Can you rebase from master, the error you're getting on CircleCI related to the Rubocop version has been fixed!

julianguyen commented on client/app/bundles/shared/components/buttonApp.jsx:
```diff
@@ -0,0 +1,150 @@
+      }
+      return (
+          <div>
+          <ButtonFilledM/>
```
Let's separate each of these components you built into separate files! You can place these files under `client/app/bundles/buttons`.

julianguyen commented on client/app/bundles/shared/components/buttonApp.jsx:
```diff
@@ -0,0 +1,150 @@
+       
+       var buttonStyle = {
+         fontSize: 18,
+         color: '#6D0839',
```
Let's create some kind of constants.jsx file with these values and use similar names as `app/assets/stylesheets/base/_values.scss` and `app/assets/stylesheets/base/_colors.scss`.

julianguyen commented on client/app/bundles/shared/components/buttonApp.jsx:
```diff
@@ -0,0 +1,150 @@
+      return (
+          <div>
+          <ButtonFilledM/>
+          <ButtonGhostM/>
```
To render these in Storybook, you'll want add to `client/app/stories/index.jsx`!

trishla08 commented:
@julianguyen hi! I have made the separate files as you can see. I don't really understand what I need to do in index.js. If you could help me out with that, I'll do that as well.

julianguyen commented on client/app/bundles/buttons/ButtonFilled.jsx:
```diff
@@ -0,0 +1,17 @@
+      return (
+          <div>
+         <button className = {css.buttonFilled}>
+            Learn More
```
This should be in our i18n files and referenced that way! Let me know if you need help figuring that out on Slack :D

julianguyen commented on client/app/bundles/buttons/ButtonFilledM.jsx:
```diff
@@ -0,0 +1,18 @@
+      return (
+          <div>
+         <button className = {css.buttonFilledM}>
+            Join
```
Same here for i18n!

julianguyen commented on client/app/bundles/buttons/ButtonGhost.jsx:
```diff
@@ -0,0 +1,18 @@
+
+class ButtonGhost extends React.Component {
+   render() {
+       
```
We don't need double spaces!

julianguyen commented on client/app/bundles/buttons/ButtonGhost.jsx:
```diff
@@ -0,0 +1,18 @@
+      return (
+          <div>
+         <button className = {css.buttonGhost}>
+            Join
```
Same here for i18n!

julianguyen commented on client/app/bundles/buttons/ButtonGhostM.jsx:
```diff
@@ -0,0 +1,17 @@
+      return (
+          <div>
+         <button className = {css.buttonGhostM}>
+            Join
```
Same here for i18n!

julianguyen commented on client/app/bundles/buttons/ButtonGhostXS.jsx:
```diff
@@ -0,0 +1,17 @@
+      return (
+          <div>
+         <button className = {css.buttonGhostXS}>
+            Join
```
Same here for i18n!

nshki commented:
@julianguyen Do you think it would make sense to just make one component that can take multiple different props? e.g:

```html
<Button filled large />
<Button small />
<Button href='https://www.if-me.org/' />
```

That could make it much easier to maintain moving forward!

trishla08 commented:
@nshki i have changed the pointer and also removed the border. 

I initially did have all those in a single file but then put them in separate ones as @julianguyen suggested..

julianguyen commented:
@nshki I actually prefer them to be separate components that way each components just focuses on the particular button style. Could you elaborate why using props to identify the button style would be easier to maintain moving forward?

nshki commented:
@julianguyen Ah, gotcha! Was thinking it could keep the code DRY-er since it would consolidate styles, but I realized that since an external Sass file is being used, consolidation can just happen in there. Also, if we wanted to create `<a>` tags with the same styles, we can just create a new component and use the same classes. Never mind me. :)

nshki commented on client/app/bundles/buttons/ButtonFilled.jsx:
```diff
@@ -0,0 +1,17 @@
+       
+       
+      return (
+          <div>
```
The `<div>` wrappers could be omitted, I think, so this component could be used inline as well!

pederjohnsen replied:
I believe you have to use `fragments` to omit the `<div>` wrappers?
https://reactjs.org/blog/2017/11/28/react-v16.2.0-fragment-support.html#jsx-fragment-syntax

baohouse replied:
@pederjohnsen Fragments is a nicer way of returning an array of HTML elements, but it's not necessary here because we're returning a single <button> element.

pederjohnsen replied:
@baohouse nice to know :) 👍

nshki commented on client/app/bundles/buttons/ButtonFilledM.jsx:
```diff
@@ -0,0 +1,18 @@
+       
+       
+      return (
+          <div>
```
Same here for `<div>`!

nshki commented on client/app/bundles/buttons/ButtonGhost.jsx:
```diff
@@ -0,0 +1,18 @@
+       
+       
+      return (
+          <div>
```
Same here for `<div>`!

nshki commented on client/app/bundles/buttons/ButtonGhostM.jsx:
```diff
@@ -0,0 +1,17 @@
+       
+      
+      return (
+          <div>
```
Same here for `<div>`!

nshki commented on client/app/bundles/buttons/ButtonGhostXS.jsx:
```diff
@@ -0,0 +1,17 @@
+       
+      
+      return (
+          <div>
```
Same here for `<div>`!

nshki commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,77 @@
+@import "base.scss";
+
+button:hover{
```
This is good, but it could be a more specific selector so that there's less of a chance its properties affect other elements! Perhaps moving the `cursor: pointer;` into each of the button classes will do.

bingo-lumity commented on client/app/stories/index.jsx:
```diff
@@ -16,6 +22,29 @@ storiesOf('Logo', module)
+.add('View', () => (
+    <ButtonFilled></ButtonFilled>
+  ));
+storiesOf('ButtonFilledM', module)
```
Just like with Logo above, the Buttons should be a single story with multiple views; keeps it clean.

baohouse replied:
Just like the Logo example above, grouping all buttons into a "Button" story would keep things simple to navigate as our component library grows.

bingo-lumity commented:
To facilitate code reusability, you may want to create a ButtonFactory function that abstracts away all the commonalities among the Buttons (e.g. adding id and onClick props).

ButtonFactory.jsx
`
export default function (variationClassName) {
  return class extends React.Component  (
    render () {
      const { className, ...props } = this.props;
      return <button {...props} className={`${className} ${variationClassName}`}/>;
    }  
  );
}
`

ButtonFilled.jsx
`
import buttonFactory from './ButtonFactory';
import css from './button.scss';

const ButtonFilled = buttonFactory(css.buttonFilled);
export default ButtonFilled;
`

baohouse commented:
To facilitate code reusability, you may want to create a ButtonFactory function that abstracts away all the commonalities among the Buttons (e.g. adding id and onClick props).

ButtonFactory.jsx
```
export default function (variationClassName) {
  return class extends React.Component {
    render () {
      const { children, className, ...props } = this.props;
      return (
        <button {...props} className={`${className} ${variationClassName}`}>
          {children}
        </button>
      );
    }  
  };
}
```

ButtonFilled.jsx
```
import buttonFactory from './ButtonFactory';
import css from './button.scss';

const ButtonFilled = buttonFactory(css.buttonFilled);
export default ButtonFilled;
```

julianguyen commented:
You can hold off the i18n changes for now. I'll have to figure out how we want to facilitate that from the client-side. I had wrongly assumed it was set up :P +1 to @nshki and @baohouse's helpful suggestions!

julianguyen commented on client/app/bundles/buttons/ButtonFilled.jsx:
```diff
@@ -0,0 +1,15 @@
+       
+      return (
+         <button className = {css.buttonFilled}>
+            Learn More
```
Let's pass the button text as props e.g. {this.props.text}

baohouse replied:
If using props, usage would look like this:
```
<ButtonFilled text="Learn More"/>
<ButtonFilled text="Learn More" icon={<Icon type="smiley"/>}/>
```
but if using children, usage would look like this
```
<ButtonFilled>
  Learn More
</ButtonFilled>

<ButtonFilled>
  <Icon type="smiley"/> Learn More
</ButtonFilled>
```

julianguyen commented on client/app/bundles/buttons/ButtonFilledM.jsx:
```diff
@@ -0,0 +1,16 @@
+       
+      return (
+         <button className = {css.buttonFilledM}>
+            Join
```
Let's pass the button text as props e.g. {this.props.text}

julianguyen commented on client/app/bundles/buttons/ButtonGhost.jsx:
```diff
@@ -0,0 +1,16 @@
+       
+      return (
+         <button className = {css.buttonGhost}>
+            Join
```
Let's pass the button text as props e.g. {this.props.text}

julianguyen commented on client/app/bundles/buttons/ButtonGhostM.jsx:
```diff
@@ -0,0 +1,15 @@
+      
+      return (
+         <button className = {css.buttonGhostM}>
+            Join
```
Let's pass the button text as props e.g. {this.props.text}

julianguyen commented on client/app/bundles/buttons/ButtonGhostXS.jsx:
```diff
@@ -0,0 +1,15 @@
+      
+      return (
+         <button className = {css.buttonGhostXS}>
+            Join
```
Let's pass the button text as props e.g. {this.props.text}

trishla08 commented:
@julianguyen @nshki the following changes have been done:
-Passed button text as props
-A little correction in CSS
-Added all buttons in a single dropdown for storybook

Let me know if anything else is needed!

nshki commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,79 @@
+         padding-bottom: 16px;
+         padding-left: 15px;
+         padding-right: 15px;
+         width: 203px;
```
Sorry for not catching this before, but let's make sure we don't give these buttons explicit dimensions! If you could keep the spacing between the text and the edges of the container consisting using padding, that would be perfect.

nshki commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,79 @@
+         font-weight : bold;
+         padding-top: 17px;
+         padding-bottom: 16px;
+         width: 128px;
```
Same here for explicit dimensions!

nshki commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,79 @@
+         padding-bottom: 22px;
+         padding-left: 15px;
+         padding-right: 15px;
+         width: 203px;
```
Same here for explicit dimensions!

nshki commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,79 @@
+         border-radius: 5px;
+         opacity: 0.95;
+         border-width: 0px;
+         width: 128px;
```
Same here for explicit dimensions!

nshki commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,79 @@
+         font-weight: bold;
+         padding-top: 17px;
+         padding-bottom: 22px;
+         width: 128px;
```
Same here for explicit dimensions!

nshki replied:
Let's make this width a `min-width` instead, and also take out the duplicate width property for this selector!

nshki commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,79 @@
+         
+}
+.buttonFilledM{
+         fontSize: 18px;
```
Should be `font-size` here!

bingo-lumity replied:
missing hyphen

nshki commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,79 @@
+         padding-bottom: 16px;
+         width: 128px;
+         border-radius: 5px;
+         opacity: 0.95;
```
Also sorry for not commenting on this before -- let's define a background color to be a 95% opacity white instead of using the opacity property here!

nshki commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,79 @@
+         padding-top: 17px;
+         padding-bottom: 22px;         
+         border-radius: 5px;
+         opacity: 0.95;
```
Same here for background color!

nshki commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,79 @@
+         width: 128px;
+         width: 149px;
+         border-radius: 5px;
+         opacity: 0.95;
```
Same here for background color!

trishla08 commented:
@nshki done but setting the background color to be rgba(255,255,255,0.95) makes it invisible on storybook since the text color is white and it might also be a browser issue for me that the opacity isn't visible properly..

trishla08 commented:
@julianguyen
@nshki

Any updates here?

nshki commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,79 @@
+         padding-left: 48px;
+         padding-right: 47px;
+         border-radius: 5px;
+         background-color: rgba(255,255,255,0.95);
```
Ah, this was my mistake with my previous suggestion! Should be 10%, not 95%, sorry!

nshki commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,79 @@
+         padding-left: 48px;
+         padding-right: 47px;
+         border-radius: 5px;
+         background-color: rgba(255,255,255,0.95);
```
Same here for 10%!

nshki commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,79 @@
+         padding-top: 17px;
+         padding-bottom: 16px;
+         padding-left: 48px;
+         padding-right: 47px;
```
Let's make sure the horizontal padding is even here. I think it will also make sense to reduce the horizontal padding to around 24px and adding a `min-width` so that it scales nicely!

nshki commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,79 @@
+         padding-top: 17px;
+         padding-bottom: 16px;
+         padding-left: 85px;
+         padding-right: 85px;
```
See comment for `.buttonGhostM`!

nshki commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,79 @@
+         padding-top: 17px;
+         padding-bottom: 22px;
+         padding-left: 85px;
+         padding-right: 85px;
```
See comment for `.buttonGhostM` but instead of 24px use 30px! :)

nshki commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,79 @@
+         padding-top: 17px;
+         padding-bottom: 22px; 
+         padding-left: 48px;
+         padding-right: 47px;
```
See comment for `.buttonGhostM` but instead of 24px use 30px! :)

nshki commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,79 @@
+         color: #FFFFFF;
+         font-weight: bold;
+         padding-top: 17px;
+         padding-bottom: 22px;
```
See comment for `.buttonGhostM` but instead of 24px use 20px!

nshki commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,79 @@
+         width: 128px;
+         width: 149px;
+         border-radius: 5px;
+         background-color: rgba(255,255,255,0.95);
```
Same here for 10%!

nshki commented:
Also, as an aside, you can create a wrapper in Storybook so that you can view the light buttons against a darker background! That will make it easier for yourself and everyone to see the buttons clearly.

trishla08 commented:
I changed the css according to what you said! @nshki

julianguyen commented on client/app/bundles/buttons/ButtonFilled.jsx:
```diff
@@ -0,0 +1,15 @@
+
+class ButtonFilled extends React.Component {
+   render() {
+       
```
Can get rid of the extra empty lines here!

julianguyen commented on client/app/bundles/buttons/button.scss:
```diff
@@ -0,0 +1,84 @@
+
+.buttonFilledM{
+         font-size: 18px;
+         color: #6D0839;
```
I would use the variables from _colors.scss here, you can import the file.

julianguyen commented:
@trishla08 How is this coming along?

trishla08 commented:
@julianguyen sorry for this delay!! Had my exams going. Which colors.scss file do I import? I can find more than 1 in the folder.

julianguyen commented:
@trishla08 No worries! I hope exams are going well :D

So we're actually already importing _colors.scss in `/Users/julia/ifme/client/app/bundles/shared/components/_base.scss` :D So you should be able to use the variables in your stylesheet. Let me know if you have any other questions

trishla08 commented:
@julianguyen colors are now imported!

trishla08 commented:
Had no idea this was not approved yet. @julianguyen @nshki has this been done by someone else?

nshki commented:
@trishla08 You're the only person working on it, from what I can see!

trishla08 commented:
@nshki do I need to maake any changes then?

baohouse commented:
@trishla08 The branch is out of sync with the master branch. You'll want to resolve the merge conflict first.

aSquare14 commented:
Please let us know If we can work on this component.
